src/Servers/Kestrel/Core/src/Internal/Http3/Http3Connection.cs

```diff
@@ -288,54 +324,87 @@ private void UpdateStreamTimeouts(DateTimeOffset now)
-                            stream.InitializeWithExistingContext(streamContext.Transport);
+                            // http request stream
+                            // https://quicwg.org/base-drafts/draft-ietf-quic-http.html#section-5.2-2
+                            if (_gracefulCloseStarted)
+                            {
+                                // https://quicwg.org/base-drafts/draft-ietf-quic-http.html#section-4.1.2-3
+                                streamContext.Features.GetRequiredFeature<IProtocolErrorCodeFeature>().Error = (long)Http3ErrorCode.RequestRejected;
+                                streamContext.Abort(new ConnectionAbortedException("HTTP/3 connection is closing and no longer accepts new requests."));
+                                await streamContext.DisposeAsync();
+
+                                return;
+                            }
+
+                            // Request stream IDs are tracked.
+                            UpdateHighestOpenedRequestStreamId(streamIdFeature.StreamId);
+
+                            var persistentStateFeature = streamContext.Features.Get<IPersistentStateFeature>();
+                            Debug.Assert(persistentStateFeature != null, $"Required {nameof(IPersistentStateFeature)} not on stream context.");
+
+                            Http3Stream stream;
+
+                            // Check whether there is an existing HTTP/3 stream on the transport stream.
+                            // A stream will only be cached if the transport stream itself is reused.
+                            if (!persistentStateFeature.State.TryGetValue(StreamPersistentStateKey, out var s))
+                            {
+                                stream = new Http3Stream<TContext>(application, context);
+                                persistentStateFeature.State.Add(StreamPersistentStateKey, stream);
+                            }
+                            else
+                            {
+                                stream = (Http3Stream<TContext>)s!;
+                                stream.InitializeWithExistingContext(streamContext.Transport);
+                            }
+
+                            _streamLifetimeHandler.OnStreamCreated(stream);
+                            KestrelEventSource.Log.RequestQueuedStart(stream, AspNetCore.Http.HttpProtocol.Http3);
+                            ThreadPool.UnsafeQueueUserWorkItem(stream, preferLocal: false);
                         }
-
-                        _streamLifetimeHandler.OnStreamCreated(stream);
-
-                        KestrelEventSource.Log.RequestQueuedStart(stream, AspNetCore.Http.HttpProtocol.Http3);
-                        ThreadPool.UnsafeQueueUserWorkItem(stream, preferLocal: false);
+                    }
+                }
+                catch (Exception ex)
+                {
+                    if (ex.Data.Contains("StreamId"))
```

Comment:
Exception data is an unusual way to pass state around. Why not a custom error with StreamId as a property?